src/lshashtable.c

```diff
@@ -0,0 +1,826 @@
+        return NULL;
+    struct lsht_record_store_rec *newrec = rs_wloc(rs);
+    _lsht_record_store_set_use(rs, _lsht_record_store_pos2blockpos(rs_wlocpos(rs)));
+    memcpy(newrec, rec, size);
+    rs_wseek(rs, size);
+    rs->num_free_blocks -= newrec->num_blocks;
+    return newrec;
+}
+
+static inline uint16_t lshashtable_get_rs_id(struct lshashtable *lsht, struct lsht_record_store *rs) {
+    struct lsht_record_store *rs_begin = (struct lsht_record_store *)vmbuf_data(&lsht->record_stores);
+    return rs - rs_begin;
+}
+
+static inline void *_next_rec(void *r) {
+    return r + (((struct lsht_record_store_rec *)r)->num_blocks << RS_BLOCK_SIZE_BITS);
+}
+
+static inline int lshashtable_is_expired(struct lshashtable *lsht, struct lshashtable_rec *rec) {
+    return rec->val_size != UINT_MAX && lsht->is_expired(rec);
+}
+
+static inline void lshashtable_make_room(struct lshashtable *lsht, struct lsht_record_store *rs, size_t size) {
+    uint64_t num_deleted_keys = 0;
+    struct vmbuf *recs = &rs->recbuf;
+    vmbuf_init(recs, 4096);
+    uint16_t rs_id = lshashtable_get_rs_id(lsht, rs), next_rs_id = rs_id + 1;
+    size_t num_free_blocks_threshold = (size >> RS_BLOCK_SIZE_BITS) + (lsht_record_store_total_blocks(rs) >> 2);
+    while (lsht_record_store_avail(rs) < size) {
+        /* reclaim space of deleted records */
+        while (!rs_empty(rs)) {
+            struct lsht_record_store_rec *r = lsht_record_store_oldest_rec(rs);
+            if (_lsht_record_store_is_inuse(rs, _lsht_record_store_pos2blockpos(rs_rlocpos(rs)))) {
+                struct lshashtable_rec *rec = (struct lshashtable_rec *)r->data;
+                /* if the record is the special record for deleted key and in the last store, we can safely discard it */
+                if (UINT_MAX != rec->val_size || next_rs_id != lsht->next_rs_id) {
+                    if (lshashtable_is_expired(lsht, rec)) {
+                        lshashtable_del_key_index(lsht, rec->data, rec->key_size, NULL);
+                        lsht_record_store_del(rs, rec);
+                    } else
+                        break;
+                }
+            }
+            rs_rseek(rs, _rec_size(r));
+        }
+        if (lsht_record_store_avail(rs) >= size)
+            break;
+        /* sanity check */
+        if (rs_empty(rs)) {
+            LOGGER_ERROR("requested size is larger than configured capacity: %zu > %zu, aborting...", size, rs_capacity(rs));
+            abort();
+        }
+        struct lsht_record_store_rec *r = lsht_record_store_pop_oldest_rec(rs);
+        if (lsht_record_store_num_free_blocks(rs) > num_free_blocks_threshold) { /* relocate */
+            size_t rec_size = _rec_size(r);
+            vmbuf_memcpy(recs, r, rec_size);
+            size += rec_size;
+        } else {/* evict or delete LRU */
+            struct lshashtable_rec *rec = (struct lshashtable_rec *)r->data;
+            if (lshashtable_is_expired(lsht, rec)) {
```

Comment:
I was wondering if we need to check expiry again here. I understand that it can happen between line 224 and 246 but we can always deal with this very unlikely case next time we do make room cleanup...